File: src/main/java/graphql/nadel/NadelExecutionInput.java
package graphql.nadel;

import graphql.PublicApi;
import graphql.execution.ExecutionId;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.ForkJoinPool;

import static graphql.GraphQLContext.newContext;
import static java.util.Objects.requireNonNull;

@PublicApi
public class NadelExecutionInput {

    private final String query;
    private final String operationName;
    private final Object context;
    private final Map<String, Object> variables;
    private final String artificialFieldsUUID;
    private final ExecutionId executionId;
    private final ForkJoinPool forkJoinPool;

    private NadelExecutionInput(String query,
                                String operationName,
                                Object context, Map<String, Object> variables,
                                String artificialFieldsUUID,
                                ExecutionId executionId,
                                ForkJoinPool forkJoinPool) {
        this.query = requireNonNull(query);
        this.operationName = operationName;
        this.context = context;
        this.variables = requireNonNull(variables);
        this.artificialFieldsUUID = artificialFieldsUUID;
        this.executionId = executionId;
        this.forkJoinPool = forkJoinPool;
    }

    public static Builder newNadelExecutionInput() {
        return new Builder();
    }

    public String getQuery() {
        return query;
    }

    public String getArtificialFieldsUUID() {
        return artificialFieldsUUID;
    }

    public String getOperationName() {
        return operationName;
    }

    public Object getContext() {
        return context;
    }

    public Map<String, Object> getVariables() {
        return new LinkedHashMap<>(variables);
    }

    public ForkJoinPool getForkJoinPool() {
        return forkJoinPool;
    }

    /**
     * @return Id that will be/was used to execute this operation.
     */
    public ExecutionId getExecutionId() {
        return executionId;
    }

    public static class Builder {
        private String query;
        private String operationName;
        private Object context = newContext().build();
        private Map<String, Object> variables = new LinkedHashMap<>();
        private String artificialFieldsUUID;
        private ExecutionId executionId;
        private ForkJoinPool forkJoinPool = ForkJoinPool.commonPool();

        private Builder() {
        }

        public Builder query(String query) {
            this.query = query;
            return this;
        }

        public Builder operationName(String operationName) {
            this.operationName = operationName;
            return this;
        }

        public Builder context(Object context) {
            this.context = context;
            return this;
        }

        public Builder variables(Map<String, Object> variables) {
            this.variables = variables;
            return this;
        }

        public Builder executionId(ExecutionId executionId) {
            this.executionId = executionId;
            return this;
        }

        public Builder artificialFieldsUUID(String artificialFieldsUUID) {
            this.artificialFieldsUUID = artificialFieldsUUID;
            return this;
        }

        public Builder forkJoinPool(ForkJoinPool forkJoinPool) {
            this.forkJoinPool = forkJoinPool;
            return this;
        }

        public NadelExecutionInput build() {
            return new NadelExecutionInput(query, operationName, context, variables, artificialFieldsUUID, executionId, forkJoinPool);
        }

    }
}

File: lib/src/main/java/graphql/nadel/NadelExecutionInput.kt
package graphql.nadel

import graphql.GraphQLContext
import graphql.execution.ExecutionId
import graphql.nadel.time.NadelInternalLatencyTracker
import graphql.nadel.time.NadelInternalLatencyTrackerImpl
import graphql.nadel.time.NadelStopwatch
import java.util.function.Consumer

data class NadelExecutionInput(
    val query: String,
    val operationName: String?,
    val context: Any?,
    val graphqlContext: GraphQLContext,
    val variables: Map<String, Any?>,
    val extensions: Map<String, Any?>,
    val executionId: ExecutionId?,
    val executionHints: NadelExecutionHints,
    val latencyTracker: NadelInternalLatencyTracker,
) {
    class Builder {
        private var query: String? = null
        private var operationName: String? = null
        private var context: Any? = null
        private var graphqlContextBuilder: GraphQLContext.Builder = GraphQLContext.newContext()
        private var variables: Map<String, Any?> = LinkedHashMap()
        private var extensions: Map<String, Any?> = LinkedHashMap()
        private var executionId: ExecutionId? = null
        private var executionHints = NadelExecutionHints.newHints().build()
        private var latencyTracker: NadelInternalLatencyTracker? = null

        fun query(query: String): Builder {
            this.query = query
            return this
        }

        fun operationName(operationName: String?): Builder {
            this.operationName = operationName
            return this
        }

        fun context(context: Any?): Builder {
            this.context = context
            return this
        }

        fun graphqlContext(contextConsumer: Consumer<GraphQLContext.Builder>): Builder {
            contextConsumer.accept(this.graphqlContextBuilder)
            return this
        }

        fun graphqlContext(key: Any, value: Any): Builder {
            this.graphqlContextBuilder.put(key, value)
            return this
        }

        fun variables(variables: Map<String, Any?>?): Builder {
            this.variables = variables ?: emptyMap()
            return this
        }

        fun extensions(extensions: Map<String, Any?>?): Builder {
            this.extensions = extensions ?: emptyMap()
            return this
        }

        fun executionId(executionId: ExecutionId): Builder {
            this.executionId = executionId
            return this
        }

        fun executionHints(nadelExecutionHints: NadelExecutionHints): Builder {
            this.executionHints = nadelExecutionHints
            return this
        }

        fun transformExecutionHints(transformFunc: (NadelExecutionHints.Builder) -> Unit): Builder {
            this.executionHints = executionHints.toBuilder()
                .also(transformFunc)
                .build()
            return this
        }

        fun latencyTracker(latencyTracker: NadelInternalLatencyTracker): Builder {
            this.latencyTracker = latencyTracker
            return this
        }

        fun build(): NadelExecutionInput {
            return NadelExecutionInput(
                query = requireNotNull(query) { "Query must be provided" },
                operationName = operationName,
                context = context,
                graphqlContext = graphqlContextBuilder.build(),
                variables = variables,
                extensions = extensions,
                executionId = executionId,
                executionHints = executionHints,
                latencyTracker = latencyTracker ?: NadelInternalLatencyTrackerImpl(NadelStopwatch()),
            )
        }
    }

    companion object {
        @JvmStatic
        fun newNadelExecutionInput(): Builder {
            return Builder()
        }
    }
}
